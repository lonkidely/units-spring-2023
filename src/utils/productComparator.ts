import type { Product, SortBy } from '../types';

// export для тестов
export const getProductRUBPrice = (product: Product) =>
    product.priceSymbol === '$' ? product.price * 70 : product.price;

export const productComparator =
    (sortBy: SortBy) =>
    (lhs: Product, rhs: Product): number => {
        if (sortBy === 'по умолчанию') return 0;

        const lhsPrice = getProductRUBPrice(lhs);
        const rhsPrice = getProductRUBPrice(rhs);

        if (rhsPrice > lhsPrice) {
            return sortBy === 'по убыванию цены' ? 1 : -1;
        }

        if (rhsPrice < lhsPrice) {
            return sortBy === 'по убыванию цены' ? -1 : 1;
        }

        return 0;
    };
