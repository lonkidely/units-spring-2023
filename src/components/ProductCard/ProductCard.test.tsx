import React from 'react';

import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { ProductCard } from './ProductCard';
import { getPrice } from '../../utils';
import { Product } from '../../types';

jest.mock('../../utils/getPrice', () => ({
    __esModule: true,
    getPrice: jest.fn(() => '300 $'),
}));

afterEach(jest.clearAllMocks);

describe('Product card test', () => {
    const testProduct: Product = {
        id: 1,
        name: 'Test product 1',
        description: 'Description of test product 1',
        price: 300,
        priceSymbol: '$',
        category: 'Для дома',
        imgUrl: '/iphone.png',
    };

    it('should render correctly with image', () => {
        const renderedCard = render(<ProductCard {...testProduct} />);
        expect(renderedCard.asFragment()).toMatchSnapshot();
        expect(renderedCard.baseElement.querySelector('.product_card_image')).toBeDefined();
    });

    it('should render correctly without image', () => {
        testProduct.imgUrl = '';
        const renderedCard = render(<ProductCard {...testProduct} />);
        expect(renderedCard.asFragment()).toMatchSnapshot();
        expect(renderedCard.baseElement.querySelector('.product_card_image')).toBeNull();
    });

    it('should call getPrice once', () => {
        expect(getPrice).toHaveBeenCalledTimes(0);
        render(<ProductCard {...testProduct} />);
        expect(getPrice).toHaveBeenCalledTimes(1);
    });
});
