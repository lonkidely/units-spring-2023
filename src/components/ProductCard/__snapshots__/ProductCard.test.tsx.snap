// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product card test should render correctly with image 1`] = `
<DocumentFragment>
  <div
    class="product-card"
  >
    <div
      class="product-card__text"
    >
      <h2
        class="product-card__name"
      >
        Test product 1
      </h2>
      <div
        class="product-card__description"
      >
        Description of test product 1
      </div>
      <h1
        class="product-card__price"
      >
        300 $
      </h1>
      <div
        class="product-card__category"
      >
        Для дома
      </div>
    </div>
    <img
      alt="Test product 1"
      class="product-card__image"
      src="/iphone.png"
    />
  </div>
</DocumentFragment>
`;

exports[`Product card test should render correctly without image 1`] = `
<DocumentFragment>
  <div
    class="product-card"
  >
    <div
      class="product-card__text"
    >
      <h2
        class="product-card__name"
      >
        Test product 1
      </h2>
      <div
        class="product-card__description"
      >
        Description of test product 1
      </div>
      <h1
        class="product-card__price"
      >
        300 $
      </h1>
      <div
        class="product-card__category"
      >
        Для дома
      </div>
    </div>
  </div>
</DocumentFragment>
`;
